lens/registry.go

```diff
@@ -0,0 +1,335 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/ipfs/go-datastore/query"
+	"github.com/lens-vm/lens/host-go/config"
+	"github.com/sourcenetwork/immutable"
+	"github.com/sourcenetwork/immutable/enumerable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+// todo: This file, particularly the `lensLocker`` stuff, contains fairly sensitive code that is both
+// cumbersome to fully test with integration/benchmark tests, and can have a significant affect on
+// the users if broken (deadlocks, large performance degradation).  It should have proper unit tests.
+// https://github.com/sourcenetwork/defradb/issues/1596
+
+// lensRegistry is responsible for managing all migration related state within a local
+// database instance.
+type lensRegistry struct {
+	lensPoolSize int
+	// lens lockers by source schema version ID
+	lensWarehouse map[string]*lensLocker
+	// lens configurations by source schema version ID
+	lensConfigs map[string]client.LensConfig
+}
+
+var _ client.LensRegistry = (*lensRegistry)(nil)
+
+// NewRegistry instantiates a new registery.
+//
+// It will be of size 5 (per schema version) if a size is not provided.
+func NewRegistry(lensPoolSize immutable.Option[int]) *lensRegistry {
+	var size int
+	if lensPoolSize.HasValue() {
+		size = lensPoolSize.Value()
+	} else {
+		size = 5
+	}
+
+	return &lensRegistry{
+		lensPoolSize:  size,
+		lensWarehouse: map[string]*lensLocker{},
+		lensConfigs:   map[string]client.LensConfig{},
+	}
+}
+
+func (r *lensRegistry) SetMigration(ctx context.Context, txn datastore.Txn, cfg client.LensConfig) error {
+	key := core.NewSchemaVersionMigrationKey(cfg.SourceSchemaVersionID)
+
+	json, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	err = txn.Systemstore().Put(ctx, key.ToDS(), json)
+	if err != nil {
+		return err
+	}
+
+	err = r.cacheLens(txn, cfg)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *lensRegistry) cacheLens(txn datastore.Txn, cfg client.LensConfig) error {
+	locker, lockerAlreadyExists := r.lensWarehouse[cfg.SourceSchemaVersionID]
+	if !lockerAlreadyExists {
+		locker = newLocker(r.lensPoolSize, cfg)
+	}
+
+	newLensPipes := make([]*lensPipe, r.lensPoolSize)
+	for i := 0; i < r.lensPoolSize; i++ {
+		var err error
+		newLensPipes[i], err = newLensPipe(cfg)
+		if err != nil {
+			return err
+		}
+	}
+
+	// todo - handling txns like this means that the migrations are not available within the current
+	// transaction if used for stuff (e.g. GQL requests) before commit.
+	// https://github.com/sourcenetwork/defradb/issues/1592
```

Comment:
I'm not sure I understand your comment.  Can you please rephrase/expand it :)

Reply:
Basically, to simplify things to elimate this problem, any action that is effecting the registry should not have a shared transaction?

Can you define the scenarios where we have some kind of DB operation that effects the registry, that may also then need the migration available to it?

Eg, if someone tried to create and run a migration in a single transaction, we should specifically not allow that imo. This can likely be done by not placing the related client methods on the `Store` interface, and instead on the `DB` interface (so it doesnt have to be exposed on `explicitDB`).

This can likely be a follow-up as its possibly a larger convo.

Reply:
>  if someone tried to create and run a migration in a single transaction

This is likely a highly desirable and common use case.  It is very very nice having the entirety of a migration (schema change(s), lens migrations, data changes, migration verification) all scoped to a single transaction and being able to do so saves a very large amount for developer time and stress that would otherwise have occurred if/when a partial failure occurs.